src/python.rs

```diff
@@ -32,11 +35,27 @@ pub fn python_parse<'py>(
     allow_inf_nan: bool,
     cache_mode: StringCacheMode,
     allow_partial: bool,
+    catch_duplicate_keys: bool,
 ) -> JsonResult<Bound<'py, PyAny>> {
+    macro_rules! ppp {
+        ($string_cache:ident, $key_check:ident) => {
+            PythonParser::<$string_cache, $key_check>::parse(py, json_data, allow_inf_nan, allow_partial)
+        };
+    }
+
     match cache_mode {
```

Comment:
@davidhewitt can we somehow use the builder pattern and thereby:
1. Avoid the proliferation of choices here
2. Avoid changing every test and benchmark when we add more settings to `python_parse`

??

Reply:
You need some promotion of runtime options to compile-time generics, so yes a builder can work to solve both problems but I think the gotcha is that the builder itself has to become generic.

e.g.

```rust
struct PythonParseOptions<StringCache, KeyCheck> {
    key_check: PhantomData<KeyCheck>
}

impl<StringCache, KeyCheck> PythonParseOptions<StringCache, KeyCheck> {
    fn duplicate_key_check(self) -> PythonParseOptions<StringCache, DuplicateKeyCheck> {
       PythonParseOptions {
           // ... spell it out here
       }
    }
}
```

If you want the builder to not be generic then I think `python_parse` necessarily needs to keep the complexity which you currently experience.

Might want to be careful about having too many generic options though, as the price we'll be paying is exponential blowup in code size.